File: src/lib/placeSearch.ts
import { normalizeSearchText } from './mediaSearch'

export interface PlaceHit {
  id: string
  label: string
  latitude: number
  longitude: number
  /** Nominatim bounding box: [south, north, west, east] */
  bbox?: [number, number, number, number]
}

interface NominatimResult {
  place_id: number
  display_name: string
  lat: string
  lon: string
  boundingbox?: [string, string, string, string]
}

/** OpenStreetMap Nominatim — yer adı → koordinat (1 req/sn nazik kullanım). */
export async function searchPlaces(
  query: string,
  signal?: AbortSignal,
): Promise<PlaceHit[]> {
  const q = query.trim()
  if (q.length < 2) return []

  const url =
    'https://nominatim.openstreetmap.org/search?' +
    new URLSearchParams({
      q,
      format: 'json',
      limit: '6',
      addressdetails: '0',
    })

  const res = await fetch(url, {
    signal,
    headers: {
      Accept: 'application/json',
    },
  })
  if (!res.ok) return []
  const data = (await res.json()) as NominatimResult[]
  const hits: PlaceHit[] = []
  for (const row of data) {
    const latitude = Number(row.lat)
    const longitude = Number(row.lon)
    if (!Number.isFinite(latitude) || !Number.isFinite(longitude)) continue
    let bbox: PlaceHit['bbox']
    if (row.boundingbox?.length === 4) {
      const south = Number(row.boundingbox[0])
      const north = Number(row.boundingbox[1])
      const west = Number(row.boundingbox[2])
      const east = Number(row.boundingbox[3])
      if (
        Number.isFinite(south) &&
        Number.isFinite(north) &&
        Number.isFinite(west) &&
        Number.isFinite(east)
      ) {
        bbox = [south, north, west, east]
      }
    }
    hits.push({
      id: `place-${row.place_id}`,
      label: row.display_name,
      latitude,
      longitude,
      bbox,
    })
  }
  return hits
}

/** Yer sonuçlarını yerel metin skoruyla sırala (normalize). */
export function rankPlaces(hits: PlaceHit[], query: string): PlaceHit[] {
  const needle = normalizeSearchText(query)
  if (!needle) return hits
  return [...hits].sort((a, b) => {
    const an = normalizeSearchText(a.label)
    const bn = normalizeSearchText(b.label)
    const as = an.startsWith(needle) ? 0 : an.includes(needle) ? 1 : 2
    const bs = bn.startsWith(needle) ? 0 : bn.includes(needle) ? 1 : 2
    return as - bs || a.label.length - b.label.length
  })
}

File: lib/services/place_search.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

class PlaceHit {
  const PlaceHit({
    required this.id,
    required this.label,
    required this.latitude,
    required this.longitude,
    this.bbox,
  });

  final String id;
  final String label;
  final double latitude;
  final double longitude;

  /// south, north, west, east
  final List<double>? bbox;
}

Future<List<PlaceHit>> searchPlaces(String query) async {
  final q = query.trim();
  if (q.length < 2) return [];
  final uri = Uri.https('nominatim.openstreetmap.org', '/search', {
    'q': q,
    'format': 'json',
    'limit': '6',
    'addressdetails': '0',
  });
  try {
    final res = await http.get(
      uri,
      headers: {
        'Accept': 'application/json',
        'User-Agent': 'MedyaAtlas/0.6',
      },
    );
    if (res.statusCode != 200) return [];
    final data = jsonDecode(res.body) as List<dynamic>;
    final hits = <PlaceHit>[];
    for (final row in data) {
      final map = Map<String, dynamic>.from(row as Map);
      final lat = double.tryParse('${map['lat']}');
      final lon = double.tryParse('${map['lon']}');
      if (lat == null ||
          lon == null ||
          !lat.isFinite ||
          !lon.isFinite ||
          lat.abs() > 90 ||
          lon.abs() > 180) {
        continue;
      }
      List<double>? bbox;
      final box = map['boundingbox'];
      if (box is List && box.length == 4) {
        final south = double.tryParse('${box[0]}');
        final north = double.tryParse('${box[1]}');
        final west = double.tryParse('${box[2]}');
        final east = double.tryParse('${box[3]}');
        if (south != null &&
            north != null &&
            west != null &&
            east != null &&
            south.isFinite &&
            north.isFinite &&
            west.isFinite &&
            east.isFinite) {
          bbox = [south, north, west, east];
        }
      }
      hits.add(
        PlaceHit(
          id: 'place-${map['place_id']}',
          label: '${map['display_name']}',
          latitude: lat,
          longitude: lon,
          bbox: bbox,
        ),
      );
    }
    return hits;
  } catch (_) {
    return [];
  }
}
